rustler_tests/native/rustler_test/src/test_atom.rs

```diff
@@ -1,28 +1,27 @@
-use rustler::types::{atom::Atom, binary::Binary};
-use rustler::{Env, NifResult, Term};
+use rustler::{Atom, Binary, Env, NifResult, Term};
 
 mod atoms {
     rustler::atoms! { ok }
 }
 
-pub fn on_load(_env: Env) {}
-
-pub fn atom_to_string<'a>(_env: Env<'a>, args: &[Term<'a>]) -> NifResult<String> {
-    args[0].atom_to_string()
+#[rustler::nif]
+pub fn atom_to_string(atom: Term) -> NifResult<String> {
+    atom.atom_to_string()
 }
 
-pub fn atom_equals_ok<'a>(_env: Env<'a>, args: &[Term<'a>]) -> bool {
-    atoms::ok() == args[0]
+#[rustler::nif]
+pub fn atom_equals_ok(atom: Atom) -> NifResult<bool> {
```

Comment:
Something is weird here and I couldn't quite figure out why, but some of these require `NifResult<T>` while others don't...

Reply:
I made an attempt to get rid of some of those `NifResult` types on a [test branch](https://github.com/evnu/rustler/tree/rustler_macros-wrap-ok).

In this example here (`atom_equals_ok`), the problem is that the expanded code first tries to decode an `Atom`, which can fail. Thus, the expanded code requires that the closure in `catch_unwind` returns a type `Result<_, rustler::error::Error>`, where the error type is defined by the decoder. Now, if the test function here returns only `bool`, we have a type mismatch. We want to return a `bool` from the closure, but we are required to return a result. We can work around this issue by always returning `Ok(nif result)` from the closure instead. This appears simple at first, but we must then ensure to remove this extraneous `Ok(_)` again when handling the result. If we do not do that, we cannot use `NifResult` anymore as a return type for a `rustler::nif` function.

@scrogson Does the explanation make sense? The diff in my test branch might be simpler to follow than this explanation :) On that particular branch, it is now possible to return bool directly from `atom_equals_ok`.

Reply:
Note that this was a bit tedious to debug. To figure out what goes wrong, I used [cargo-expand](https://github.com/dtolnay/cargo-expand) on the current version of the code and replaced the code for `atom_equals_ok` with the expanded version. Then, I hacked and slashed around in the expanded code until I was able to reproduce the error, and then played around with wrapping the result in `Ok`. Alas, `cargo-expand` only works if the code compiles, so there is some back-and-forth needed between fixing something in codegen and debugging it in the expanded code.

Reply:
Yeah, I had tried a few similar things but I think the missing link for me was specifying the actual return type in the `catch_unwind`. I finally just settled to post a PR and hopefully figure it out later. It was nearly 2am 😆